### PR TITLE
Channel receive instrumentation

### PR DESCRIPTION
**Pull request by tokafish**
See https://groups.google.com/forum/#!topic/phoenix-core/SIHXEWEoFbA

Example output:

> [info] RECEIVED message on MyApp.ApiChannel
>   Parameters: %{"example" => "Hi"}
>   Ref: 20784513-6CF4-441D-9807-17E10B8D175E

### CODE REVIEW

**Comment by chrismccord** [comment]
Thanks, but we decided against logging/instrumenting every incoming event out of the box. If you need such a setup you can define a single `handle_in` clause that logs + instruments and delegates the event to a private function on the module, such as `defp incoming("event1", params, socket)`

**Comment by tokafish** [comment]
One thing I noticed in putting this together: It looks like `Phoenix.Endpoint.Instrument` adds `Phoenix.Logger` as an instrumenter, regardless of the app's config. Should the phoenix.new generator create a config with `Phoenix.Logger` explicitly included, giving the user the option to remove it?

**Comment by tokafish** [comment]
@chrismccord Thanks. @josevalim seemed to suggest that you might be open to this in the link above, which is why I submitted it. Might I ask, if there was no runtime cost to adding this instrumentation (for example, it could be turned off based on the configured log level), would you reconsider this functionality?

**Comment by josevalim** [comment]
@chrismccord, we decided to wait until someone needed it. I guess the time has arrived? :)

I think we should instrument this one because it is impossible for the users to do so (opposite to calling broadcast) and if nobody is "listening", it should be no-cost, right?

**Comment by josevalim** [comment]
To be clear, I think we should merge but without the Phoenix.Logger instrumentation for it.

**Comment by chrismccord** [comment]
It's possible by define a catch-all handle_in yourself and wrapping in instrument, but I can be onboard with instrumenting the event. @tokafish can you remove the logger bit?

**Comment by tokafish** [comment]
@chrismccord @josevalim Thanks much! Removed the logger bit and added a tiny bit of documentation. Let me know if there's anything else I can do.

**Comment by tokafish** [comment]
Also, I'm curious if anyone has a suggestion for instrumenting broadcast/push within an app. Since using Phoenix.Channel imports it, I can't define functions named `push` and `broadcast`. I guess the functions could be named differently, or called within an unimported module, but it just feels (slightly) awkward to me.

**Comment by josevalim** [comment]
You can even unimport them in your App.Web and bring the ones you want.

**Comment by tokafish** [comment]
That would be awesome! Forgive my ignorance, how do I unimport? :)

**Comment by tokafish** [comment]
For anyone who's curious to the solution, simply importing Phoenix.Channel again, with an `except` clause, was how to unimport a function 😄

File: lib/phoenix/endpoint.ex [hunk 1]
```diff
@@ -345,6 +345,8 @@ defmodule Phoenix.Endpoint do
       `"index.html"` - and the `:format` key - for the format of the template.
     * `:phoenix_channel_join` - the joining of a channel. The `%Phoenix.Socket{}`
       and join params are passed as runtime metadata via `:socket` and `:params`.
+    * `:phoenix_channel_receive` - the receiving of a message oover a channel. The `%Phoenix.Socket{}`,
```
Comment (jquadrin) [inline lib/phoenix/endpoint.ex]: oover

Comment (tokafish) [inline lib/phoenix/endpoint.ex]: @jquadrin thanks!

**Comment by tokafish** [comment]
@chrismccord Just wanted to check in and see if there's anything I should do to help me this forward. Thanks!

**Comment by chrismccord** [comment]
<3 <3 <3

**Pull request merged**